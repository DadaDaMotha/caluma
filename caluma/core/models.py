import uuid

from django.db import models


class BaseModel(models.Model):
    created_at = models.DateTimeField(auto_now_add=True)
    modified_at = models.DateTimeField(auto_now=True)
    created_by_user = models.CharField(max_length=150, blank=True, null=True)
    created_by_group = models.CharField(
        max_length=150, blank=True, null=True, db_index=True
    )

    class Meta:
        abstract = True


class SlugModel(BaseModel):
    """
    Models which use a slug as primary key.

    Defined as Caluma default for configuration so it is possible
    to merge between developer and user configuration.
    """

    slug = models.SlugField(max_length=50, primary_key=True)

    def __str__(self):
        return self.slug

    class Meta:
        abstract = True


class UUIDModel(BaseModel):
    """
    Models which use uuid as primary key.

    Defined as Caluma default
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)

    class Meta:
        abstract = True


class NaturalKeyModel(BaseModel):
    """Models which use a natural key as primary key."""

    id = models.CharField(max_length=255, unique=True, primary_key=True)

    def natural_key(self):  # pragma: no cover
        raise NotImplementedError()

    def save(self, *args, **kwargs):
        self.id = self.natural_key()
        return super().save(*args, **kwargs)

    class Meta:
        abstract = True
