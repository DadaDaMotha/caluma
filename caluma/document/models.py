from django.contrib.postgres.fields import JSONField
from django.db import models

from caluma.models import UUIDModel


class Document(UUIDModel):
    # TODO: add user field once authentication is implemented
    form = models.ForeignKey(
        "form.Form", on_delete=models.DO_NOTHING, related_name="documents"
    )
    meta = JSONField(default={})


class Answer(UUIDModel):
    question = models.ForeignKey(
        "form.Question", on_delete=models.DO_NOTHING, related_name="answers"
    )
    value = JSONField()
    meta = JSONField(default={})
    document = models.ForeignKey(
        Document, on_delete=models.CASCADE, related_name="answers"
    )

    class Meta:
        # a question may only be answerd once per document
        unique_together = ("document", "question")
