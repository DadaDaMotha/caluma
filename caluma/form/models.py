from django.contrib.postgres.fields import JSONField
from django.db import models


class Form(models.Model):
    slug = models.SlugField(max_length=50, primary_key=True)
    name = models.CharField(max_length=255)
    description = models.TextField()
    meta = JSONField()
