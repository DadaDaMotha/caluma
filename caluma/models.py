from django.db import models


class BaseModel(models.Model):
    created = models.DateTimeField(auto_now_add=True)
    modified = models.DateTimeField(auto_now=True)

    class Meta:
        abstract = True


class SlugModel(BaseModel):
    """Models which use a slug as primary key."""

    slug = models.SlugField(max_length=50, primary_key=True)

    def __str__(self):
        return self.slug

    class Meta:
        abstract = True
