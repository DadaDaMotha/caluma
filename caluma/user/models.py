from django.conf import settings


class AnonymousUser(object):
    def __init__(self):
        self.username = None
        self.groups = []
        self.token = None
        self.decoded_token = {}
        self.group = None

    @property
    def is_authenticated(self):
        return False

    def __str__(self):
        return "AnonymousUser"


class OIDCUser(object):
    def __init__(self, token, decoded_token):
        self.token = token
        self.username = decoded_token["sub"]
        self.decoded_token = decoded_token
        self.groups = decoded_token.get(settings.OIDC_GROUPS_CLAIM) or []

    @property
    def group(self):
        return self.groups and self.groups[0]

    @property
    def is_authenticated(self):
        return True

    def __str__(self):
        return self.username
