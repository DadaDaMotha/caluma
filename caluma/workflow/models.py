from django.contrib.postgres.fields import JSONField
from django.db import models
from localized_fields.fields import LocalizedField

from caluma.models import BaseModel, SlugModel


class TaskSpecification(SlugModel):
    TYPE_SIMPLE = "simple"

    TYPE_CHOICES = (TYPE_SIMPLE,)
    TYPE_CHOICES_TUPLE = ((TYPE_SIMPLE, "Task which can only be marked as completed."),)

    name = LocalizedField(blank=False, null=False, required=False)
    description = LocalizedField(blank=True, null=True, required=False)
    type = models.CharField(choices=TYPE_CHOICES_TUPLE, max_length=50)
    meta = JSONField(default={})
    is_archived = models.BooleanField(default=False)


class WorkflowSpecification(SlugModel):
    name = LocalizedField(blank=False, null=False, required=False)
    description = LocalizedField(blank=True, null=True, required=False)
    meta = JSONField(default={})
    is_published = models.BooleanField(default=False)
    is_archived = models.BooleanField(default=False)
    start = models.ForeignKey(
        TaskSpecification, on_delete=models.CASCADE, related_name="+"
    )


class Flow(BaseModel):
    workflow_specification = models.ForeignKey(
        WorkflowSpecification, related_name="flows"
    )
    task_specification = models.ForeignKey(TaskSpecification, related_name="flows")
    next = models.TextField()

    class Meta:
        unique_together = ("workflow_specification", "task_specification")
