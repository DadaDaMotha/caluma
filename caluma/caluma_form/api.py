from typing import Any, Optional

from caluma.caluma_form import domain_logic, models
from caluma.caluma_user.models import BaseUser


def save_answer(
    question: models.Question,
    document: models.Document,
    user: Optional[BaseUser] = None,
    value: Optional[Any] = None,
    context: Optional[dict] = None,
    **kwargs
) -> models.Answer:
    """
    Save an answer for given question, document.

    Similar to saveDocumentStringAnswer and the likes, it performes upsert.
    :param value: Must match the question type
    """

    data = {"question": question, "document": document, "value": value}
    data.update(kwargs)

    answer = models.Answer.objects.filter(question=question, document=document).first()

    validated_data = domain_logic.SaveAnswerLogic.pre_save(
        domain_logic.SaveAnswerLogic.validate_for_save(
            data, user, answer=answer, origin=True
        )
    )

    if answer is None:
        answer = domain_logic.SaveAnswerLogic.create(validated_data, user)
    else:
        answer = domain_logic.SaveAnswerLogic.update(validated_data, answer)

    domain_logic.SaveAnswerLogic.post_save(answer)

    return answer
