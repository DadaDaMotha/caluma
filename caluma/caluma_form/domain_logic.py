from django.db import transaction

from caluma.caluma_form import models, validators
from caluma.caluma_user.models import BaseUser
from caluma.utils import update_model


class SaveAnswerLogic:
    @staticmethod
    def validate_for_save(
        data: dict, user: BaseUser, answer: models.Answer = None, origin: bool = False
    ) -> dict:
        question = data["question"]

        if question.type == models.Question.TYPE_TABLE:
            data["documents"] = models.Document.objects.filter(pk__in=data["value"])
            del data["value"]
        elif question.type == models.Question.TYPE_FILE:
            data["file"] = data["value"]
            del data["value"]
        elif question.type == models.Question.TYPE_DATE:
            data["date"] = data["value"]
            del data["value"]

        validators.AnswerValidator().validate(
            **data, user=user, instance=answer, origin=origin
        )

        return data

    @staticmethod
    def pre_save(validated_data: dict) -> dict:
        # TODO emit events
        return validated_data

    @staticmethod
    def post_save(answer: models.Answer) -> models.Answer:
        # TODO emit events
        return answer

    @staticmethod
    @transaction.atomic
    def create(validated_data, user):
        if validated_data["question"].type == models.Question.TYPE_FILE:
            validated_data = __class__.set_file(validated_data)

        if validated_data["question"].type == models.Question.TYPE_TABLE:
            documents = validated_data.pop("documents")

        if user:
            validated_data["created_by_user"] = user
            validated_data["created_by_group"] = user.group

        answer = models.Answer.objects.create(**validated_data)

        if answer.question.type == models.Question.TYPE_TABLE:
            answer.create_answer_documents(documents)

        return answer

    @staticmethod
    @transaction.atomic
    def update(validated_data, answer):
        if answer.question.type == models.Question.TYPE_TABLE:
            documents = validated_data.pop("documents")
            answer.unlink_unused_rows(docs_to_keep=documents)

        if (
            answer.question.type == models.Question.TYPE_FILE
            and answer.file.name is not validated_data["file"]
        ):
            answer.file.delete()
            validated_data = __class__.set_file(validated_data)

        update_model(answer, validated_data)

        if answer.question.type == models.Question.TYPE_TABLE:
            answer.create_answer_documents(documents)

    @staticmethod
    def set_file(validated_data):
        file_name = validated_data.get("file")
        file = models.File.objects.create(name=file_name)
        validated_data["file"] = file
        return validated_data
