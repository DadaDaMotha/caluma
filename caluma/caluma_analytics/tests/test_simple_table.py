import math
import random

import pytest

from caluma.caluma_analytics.models import AnalyticsField
from caluma.caluma_analytics.pivot_table import PivotTable
from caluma.caluma_analytics.simple_table import SimpleTable
from caluma.caluma_form.models import Form


@pytest.mark.parametrize("analytics_table__starting_object", ["cases"])
@pytest.mark.freeze_time("2021-10-10")
def test_run_analytics_direct(db, snapshot, example_analytics, analytics_cases):
    """Test basic analytics run on simple table.

    * Test that we get *some* results
    * Tests date part extraction
    * Test Meta field extraction
    """

    table = SimpleTable(example_analytics)

    result = table.get_records()

    assert len(result) == 10

    snapshot.assert_match(result)


@pytest.mark.parametrize("analytics_table__starting_object", ["cases"])
@pytest.mark.freeze_time("2021-10-10")
@pytest.mark.parametrize("delete_case", [True, False])
@pytest.mark.parametrize("case__meta", [{"foo": "bar"}])
def test_run_analytics_gql(
    db, snapshot, schema_executor, example_analytics, case, delete_case
):
    """Test basic analytics run on simple table.

    * Test that we get *some* results
    * Tests date part extraction
    * Test Meta field extraction
    """
    query = """
        query run ($input: String!) {
          analyticsTable(slug: $input) {
            resultData {
              records {
                edges {
                  node {
                    edges {
                      node {
                        alias
                        value
                      }
                    }
                  }
                }
              }
              summary {
                edges {
                  node {
                    alias
                    value
                  }
                }
              }
            }
          }
        }
    """

    example_analytics.fields.filter(alias="from_the_doc").delete()
    example_analytics.fields.filter(alias="sub_question_sumsumsum").delete()

    if delete_case:
        # Test empty output
        case.delete()

    result = schema_executor(query, variable_values={"input": example_analytics.pk})

    assert not result.errors

    if not delete_case:
        # some explicit sanity checks
        records = result.data["analyticsTable"]["resultData"]["records"]["edges"]
        assert len(records) == 1
        first_row = {
            edge["node"]["alias"]: edge["node"]["value"]
            for edge in records[0]["node"]["edges"]
        }

        assert first_row["quarter"] == str(int(math.ceil(case.created_at.month / 3)))
        assert first_row["foo"] == "bar"

    snapshot.assert_match(result.data)


def test_sql_repeatability(
    db,
    example_analytics,
):

    example_analytics.fields.filter(alias="from_the_doc").delete()
    example_analytics.fields.filter(alias="sub_question_sumsumsum").delete()
    table = SimpleTable(example_analytics)

    sql1, params1 = table.get_sql_and_params()
    sql2, params2 = table.get_sql_and_params()

    assert sql1 == sql2
    assert params1 == params2


@pytest.mark.parametrize(
    "alias",
    [
        "hello-world",
        "with spaces",
        "UpperCase",
        "with punctuation.",
        "with punctuation?",
        "with punctuation!",
        "Lorem ipsum dolor sit amet, consetetur sadipsci",  # will be md5-ed
    ],
)
@pytest.mark.parametrize(
    "table",
    [
        "example_analytics",
        "example_pivot_table",
    ],
)
@pytest.mark.freeze_time("2021-10-10")
def test_unusual_aliases(db, table, analytics_cases, alias, request):

    table_obj = request.getfixturevalue(table)

    some_field = table_obj.fields.get(alias="quarter")
    some_field.alias = alias
    some_field.function = "sum"
    some_field.save()

    table = (
        SimpleTable(table_obj) if table_obj.is_extraction() else PivotTable(table_obj)
    )

    result = table.get_records()

    # We just check that the anlysis run went successful
    # and that the alias is represented in the columns
    assert result
    assert alias in result[0]


@pytest.mark.parametrize("use_lang", [None, "de", "en"])
def test_extract_choice_labels(
    settings,
    db,
    form_question_factory,
    question_option_factory,
    answer_factory,
    example_analytics,
    analytics_cases,
    use_lang,
    form_and_document,
):
    settings.LANGUAGES = [("de", "de"), ("en", "en")]

    # We need a form with a choice question ...
    form, *_ = form_and_document(False, False)
    choice_q = form_question_factory(form=form, question__type="choice").question
    options = question_option_factory.create_batch(4, question=choice_q)

    # ... as well as some cases with corresponding answers in their docs
    for case in analytics_cases:
        answer_factory(
            question=choice_q,
            document=case.document,
            value=random.choice(options).option.slug,
        )

        # just checking assumptions..
        assert case.document.form_id == form.pk

    # we're only interested in seeing the choice field work here
    example_analytics.fields.all().delete()
    lang_suffix = f".{use_lang}" if use_lang else ""
    example_analytics.fields.create(
        data_source=f"document[top_form].{choice_q.slug}",
        alias="choice_value",
    )
    example_analytics.fields.create(
        data_source=f"document[top_form].{choice_q.slug}.label{lang_suffix}",
        alias="choice_label",
    )

    # Define valid outputs for label and value (slug)
    lang = use_lang if use_lang else settings.LANGUAGE_CODE
    valid_options = {opt.slug: opt.label[lang] for opt in choice_q.options.all()}

    # Run the analysis
    table = SimpleTable(example_analytics)
    result = table.get_records()

    choice_count = 0
    assert len(result)
    for row in result:
        if row["choice_value"]:
            choice_count += 1
            assert row["choice_value"] in valid_options
            assert row["choice_label"] == valid_options[row["choice_value"]]
    assert choice_count >= len(analytics_cases)


@pytest.mark.parametrize(
    "analytics_table__starting_object, data_source",
    [
        ("cases", "document[*].caluma_form.name"),
        ("documents", "caluma_form.name"),
    ],
)
def test_form_info_field_cases(db, analytics_table, analytics_cases, data_source):
    valid_form_names = [
        str(n) for n in Form.objects.all().values_list("name", flat=True)
    ]

    analytics_table.fields.create(
        alias="the_thing",
        data_source=data_source,
        function=AnalyticsField.FUNCTION_VALUE,
    )

    table = SimpleTable(analytics_table)

    records = table.get_records()
    assert records
    for record in records:
        assert record["the_thing"] in valid_form_names
