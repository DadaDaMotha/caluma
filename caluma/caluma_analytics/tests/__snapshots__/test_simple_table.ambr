# name: test_run_analytics_direct[cases]
  list([
    dict({
      'created_at': datetime.datetime(2022, 2, 1, 0, 0, tzinfo=<UTC>),
      'foo': 'bar',
      'from_the_doc': 'Annette Bradley',
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': '9193',
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 1, 0, 0, tzinfo=<UTC>),
      'foo': None,
      'from_the_doc': None,
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': None,
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 2, 0, 0, tzinfo=<UTC>),
      'foo': 'bar',
      'from_the_doc': 'Brendan Anthony',
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': '5765',
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 2, 0, 0, tzinfo=<UTC>),
      'foo': None,
      'from_the_doc': None,
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': None,
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 3, 0, 0, tzinfo=<UTC>),
      'foo': 'bar',
      'from_the_doc': 'Susan Valenzuela',
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': '8625',
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 3, 0, 0, tzinfo=<UTC>),
      'foo': None,
      'from_the_doc': None,
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': None,
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 4, 0, 0, tzinfo=<UTC>),
      'foo': 'bar',
      'from_the_doc': 'Michelle Wise',
      'quarter': 1,
      'statuuuuuus': 'completed',
      'sub_question_sumsumsum': '7437',
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 4, 0, 0, tzinfo=<UTC>),
      'foo': None,
      'from_the_doc': None,
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': None,
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 5, 0, 0, tzinfo=<UTC>),
      'foo': 'bar',
      'from_the_doc': 'Nicholas Clarke',
      'quarter': 1,
      'statuuuuuus': 'suspended',
      'sub_question_sumsumsum': '7881',
    }),
    dict({
      'created_at': datetime.datetime(2022, 2, 5, 0, 0, tzinfo=<UTC>),
      'foo': None,
      'from_the_doc': None,
      'quarter': 1,
      'statuuuuuus': 'running',
      'sub_question_sumsumsum': None,
    }),
  ])
# ---
# name: test_run_analytics_gql[case__meta0-False-cases]
  dict({
    'analyticsTable': dict({
      'resultData': dict({
        'records': dict({
          'edges': list([
            dict({
              'node': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'alias': 'created_at',
                      'value': '2021-10-10 00:00:00+00:00',
                    }),
                  }),
                  dict({
                    'node': dict({
                      'alias': 'foo',
                      'value': 'bar',
                    }),
                  }),
                  dict({
                    'node': dict({
                      'alias': 'quarter',
                      'value': '4',
                    }),
                  }),
                  dict({
                    'node': dict({
                      'alias': 'statuuuuuus',
                      'value': 'running',
                    }),
                  }),
                ]),
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_run_analytics_gql[case__meta0-True-cases]
  dict({
    'analyticsTable': dict({
      'resultData': dict({
        'records': dict({
          'edges': list([
          ]),
        }),
      }),
    }),
  })
# ---
